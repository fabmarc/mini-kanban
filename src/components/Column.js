import React, { useState, useEffect } from 'react';
import styled, { css } from 'styled-components';
import { useDrop } from 'react-dnd';
import PropTypes from 'prop-types';
import _ from 'lodash';

import TextBox from './TextBox';
import Ticket from './Ticket';
import Button from './Button';
import Title from './Title';

const Wrapper = styled.div`
  width: ${props => props.width || '100%'};
  height: 100%;
  border-left: 1px solid #c0c0c0;
  &:first-child {
    border-left: 0;
  }
  ${props => props.highlight &&
    css`
      box-shadow: 0 0 8px 8px #d4e6f1;
    `}
`;

const Header = styled.div`
  width: 100%;
  height: 58px;
  display: flex;
  align-items: center;
`;

const Tickets = styled.div`
  height: calc(100% - 58px);
  padding: 16px;
`;

const inputRef = React.createRef();

function Column({ header, status, width, tickets, onChange, onDrop }) {
  const [innerTickets, setInnerTickets] = useState(tickets);
  const [description, setDescription] = useState('');
  const [newTicket, setNewTicket] = useState(false);

  useEffect(() => {
    if (newTicket && inputRef.current) inputRef.current.focus();
  }, [newTicket]);

  useEffect(() => { setInnerTickets(tickets) }, [tickets]);

  useEffect(() => {
    if (onChange) onChange(status, innerTickets);
  }, [status, onChange, innerTickets]);

  const [{ isOver, canDrop }, drop] = useDrop({
    accept: 'ticket',
    drop({ ticket }) {
      onDrop(status, ticket);
    },
    canDrop({ ticket }) {
      return (status !== ticket.status);
    },
    collect: monitor => ({
      isOver: !!monitor.isOver(),
      canDrop: !!monitor.canDrop(),
    }),
  });

  const createTicket = () => {
    if (description) {
      setInnerTickets((prevInnerTickets) => {
        const id = new Date().getUTCMilliseconds();
        return {
          ...prevInnerTickets,
          [id]: { id, description, status },
        };
      });
    }
    setDescription('');
    setNewTicket(false);
  }

  const handleValueChange = (event) => {
    setDescription(event.target.value);
  }

  const handleTicketChange = (ticket, newDescription) => {
    setInnerTickets((prevInnerTickets) => {
      prevInnerTickets[ticket.id] = {
        ...prevInnerTickets[ticket.id],
        description: newDescription,
      };
      return { ...prevInnerTickets };
    });
  }

  const handleTicketDelete = (ticket) => {
    setInnerTickets((prevInnerTickets) => {
      delete prevInnerTickets[ticket.id];
      return { ...prevInnerTickets };
    });
  }

  const handleAddClick = () => {
    setDescription('');
    setNewTicket(true);
  }

  const ticketIds = Object.keys(innerTickets || {});

  return (
    <Wrapper
      ref={drop}
      width={width}
      highlight={isOver && canDrop}
    >
      <Header>
        <Title bold>{header}</Title>
        <Button
          background="#fffef0"
          onClick={handleAddClick}
        >+</Button>
      </Header>
      <Tickets>
        {
          newTicket &&
          <TextBox
            ref={inputRef}
            placeholder="New item..."
            onBlur={createTicket}
            value={description}
            onChange={handleValueChange}
          />
        }
        {_.map(ticketIds, ticketId => {
          return (
            <Ticket
              key={ticketId}
              value={innerTickets[ticketId]}
              onChange={handleTicketChange}
              onDelete={handleTicketDelete}
            />
          );
        })}
      </Tickets>
    </Wrapper>
  );
}

Column.propTypes = {
  header: PropTypes.string.isRequired,
  status: PropTypes.string.isRequired,
  tickets: PropTypes.objectOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      description: PropTypes.string.isRequired,
      status: PropTypes.string.isRequired,
    })),
  width: PropTypes.string,
  onChange: PropTypes.func,
  onDrop: PropTypes.func,
};

export default Column;
