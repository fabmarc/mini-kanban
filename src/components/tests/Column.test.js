import React from 'react';
import { shallow, mount } from 'enzyme';
import HTML5Backend from 'react-dnd-html5-backend';
import renderer from 'react-test-renderer';
import { DndProvider } from 'react-dnd';
import Column from '../Column';

describe('<Column>', () => {
  it('renders without crashing', () => {
    shallow(
      <DndProvider backend={HTML5Backend}>
        <Column header="To-Do" status="todo" />
      </DndProvider>
    );
  });

  it('renders in edit mode', () => {
    const wrapper = mount(
      <DndProvider backend={HTML5Backend}>
        <Column header="To-Do" status="todo" />
      </DndProvider>
    );
    wrapper.find('div button').first().simulate('click');
    expect(wrapper.find('div div input').length).toEqual(1);
  });

  it('renders in edit mode', () => {
    const wrapper = mount(
      <DndProvider backend={HTML5Backend}>
        <Column header="To-Do" status="todo" />
      </DndProvider>
    );
    wrapper.find('div button').first().simulate('click');
    expect(wrapper.find('div div input').length).toEqual(1);
  });

  it('renders a just created ticket', () => {
    const wrapper = mount(
      <DndProvider backend={HTML5Backend}>
        <Column header="To-Do" status="todo" />
      </DndProvider>
    );
    wrapper.find('div button').first().simulate('click');
    wrapper.find('div div input').simulate('change', { target: { value: 'Test' } });
    wrapper.find('div div input').simulate('blur');
    expect(wrapper.find('div div div').contains('Test')).toEqual(true);
  });

  it('triggers onChange event', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <DndProvider backend={HTML5Backend}>
        <Column header="To-Do" status="todo" onChange={onChange} />
      </DndProvider>
    );
    wrapper.find('div button').first().simulate('click');
    wrapper.find('div div input').simulate('change', { target: { value: 'Test' } });
    wrapper.find('div div input').simulate('blur');
    expect(onChange.mock.calls.length).toEqual(2);
  });


  it('renders correctly with no tickets', () => {
    const tree = renderer
      .create(
        <DndProvider backend={HTML5Backend}>
          <Column header="To-Do" status="todo" />
        </DndProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with tickets', () => {
    const tree = renderer
      .create(
        <DndProvider backend={HTML5Backend}>
          <Column header="To-Do" status="todo"
            tickets={{
              1: { id: 1, description: 'Test 1', status: 'todo' },
              2: { id: 2, description: 'Test 2', status: 'todo' },
            }} />
        </DndProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
