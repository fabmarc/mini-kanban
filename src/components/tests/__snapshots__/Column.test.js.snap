// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Column> renders correctly with no tickets 1`] = `
<div
  className="sc-ifAKCX itMQZt"
>
  <div
    className="sc-EHOje kYqLhJ"
  >
    <div
      className="sc-htpNat cXCOUR"
    >
      To-Do
    </div>
    <button
      className="sc-bwzfXH kODISU"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div
    className="sc-bZQynM lfIEnT"
  />
</div>
`;

exports[`<Column> renders correctly with tickets 1`] = `
<div
  className="sc-ifAKCX itMQZt"
>
  <div
    className="sc-EHOje kYqLhJ"
  >
    <div
      className="sc-htpNat cXCOUR"
    >
      To-Do
    </div>
    <button
      className="sc-bwzfXH kODISU"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div
    className="sc-bZQynM lfIEnT"
  >
    <div
      className="sc-bxivhb gGKUWd"
      onDoubleClick={[Function]}
      opacity={1}
    >
      <div
        className="sc-htpNat cmbrzm"
      >
        Test 1
      </div>
      <button
        className="sc-bwzfXH eUNjes"
        onClick={[Function]}
      >
        -
      </button>
    </div>
    <div
      className="sc-bxivhb gGKUWd"
      onDoubleClick={[Function]}
      opacity={1}
    >
      <div
        className="sc-htpNat cmbrzm"
      >
        Test 2
      </div>
      <button
        className="sc-bwzfXH eUNjes"
        onClick={[Function]}
      >
        -
      </button>
    </div>
  </div>
</div>
`;
